src/runtime/ui-particle-base.ts

```diff
@@ -123,124 +95,97 @@ export class UiParticleBase extends Particle {
   }
 
   /**
-   * Remove all entities from named handle.
+   * invoke async function `task` with Particle busy-guard
    */
-  async clearHandle(handleName: string): Promise<void> {
-    const handle = this.handles.get(handleName);
-    if (handle instanceof Singleton || handle instanceof Collection) {
-      await handle.clear();
-    } else {
-      throw new Error('Singleton/Collection required');
-    }
+  async await(task: (p: this) => Promise<any>) {
+    return await this.invokeSafely(task, err => { throw(err); });
   }
 
   /**
-   * Merge entities from Array into named handle.
+   * Set handle value. Value can be an Entity or a POJO (or an Array of such values, for a Collection)
    */
-  async mergeEntitiesToHandle(handleName: string, entities: Entity[]): Promise<void> {
-    const idMap = {};
+  async set(handleName, value: Entity | {} | [Entity] | [{}]): Promise<void> {
+    await this.await(p => p._set(handleName, value));
+  }
+  private async _set(handleName, value: Entity | {} | [Entity] | [{}]): Promise<void> {
     const handle = this.handles.get(handleName);
-    if (handle instanceof Collection) {
-      const handleEntities = await handle.toList();
-      handleEntities.forEach(entity => idMap[entity.id] = entity);
-      for (const entity of entities) {
-        if (!idMap[this.idFor(entity)]) {
-          await handle.store(entity);
+    if (!handle) {
+      throw new Error(`Could not find handle [${handleName}]`);
+    } else {
+      if (handle instanceof Singleton) {
+        if (Array.isArray(value)) {
+          throw new Error(`Cannot set an Array to Singleton handle [${handleName}]`);
         }
+        const entity = (value instanceof Entity) ? value : new (handle.entityClass)(value);
+        return await handle['set'](entity);
       }
-    } else {
-      throw new Error('Collection required');
-    }
-  }
-
-  /**
-   * Append entities from Array to named handle.
-   */
-  async appendEntitiesToHandle(handleName: string, entities: Entity[]): Promise<void> {
-    const handle = this.handles.get(handleName);
-    if (handle) {
-      if (handle instanceof Collection || handle instanceof BigCollection) {
-        await Promise.all(entities.map(entity => handle.store(entity)));
-      } else {
-        throw new Error('Collection required');
+      else if (handle instanceof Collection) {
+        await this.clear(name);
```

Comment:
Should `set` only be for singletons? This feels like it could lead to bugs with users accidentally using set on a collection instead of add.

Reply:
Set is intended to work for Collection too.

Reply:
I disallowed calling set on Collection, to avoid confusion between set and add. If one really intends to replace the entire collection, call clear first.

Reply:
:+1: